302: Prevent a PHP fatal error on import

SergeyBiryukov opened the pull request:
Fixes #302.

SergeyBiryukov committed:
Check `GP::$original->create()` result in `GP_Original::import_for_project()`

SergeyBiryukov committed:
On import, make sure context fits into VARCHAR(255)

SergeyBiryukov commented:
The first commit prevents the fatal error from #302 and makes `GP_Original::import_for_project()` return the number of errors encountered on import.

The second one trims the context field to 255 characters on import (to match the database schema). I guess it won't cause the iOS app to display the string untranslated, as the concept of context does not exist there anyway, per [#glotpress421](https://glotpress.trac.wordpress.org/ticket/421).

toolstack commented on gp-includes/things/original.php:
```diff
@@ -107,8 +107,15 @@ public function import_for_project( $project, $translations ) {
 		$possibly_added = $possibly_dropped = array();
 
-		foreach( $translations->entries as $entry ) {
+		foreach( $translations->entries as $key => $entry ) {
```
Add a space after the `foreach` keyword?

SergeyBiryukov replied:
Done.

toolstack commented on gp-includes/things/original.php:
```diff
@@ -107,8 +107,15 @@ public function import_for_project( $project, $translations ) {
+			// Context needs to match VARCHAR(255) in the database schema.
+			if ( gp_strlen( $entry->context ) > 255 ) {
+				$entry->context = gp_substr( $entry->context, 0, 255 );
+				$translations->entries[ $entry->key() ] = $entry;
```
Instead of re-assigning the entire `$entry` should it just assign the context? ie:

`$translations->entries[ $entry->key() ]->context = $entry->context;`

Just so we don't change the array that we're already in the middle of looping through?

SergeyBiryukov replied:
We do need to assign the entire `$entry` here, because trimming `$entry->context` changes `$entry->key()`, and the `array_key_exists( $key, $translations->entries )` check fails later.

Without this assignment, the affected string ends up in `$possibly_dropped` and gets repeatedly removed / re-added on each subsequent import.

toolstack replied:
Aren't we then creating 'duplicate' entries?  Do we need to delete the old entry from the array?

ocean90 replied:
Thought about this a bit and I think can't remove the other entry or stop the processing because otherwise an existing original doesn't get updated.

ocean90 replied:
> We do need to assign the entire `$entry` here, because trimming `$entry->context` changes `$entry->key()`

For some reasons `$entry->key()` remains the same here.

```
.array(2) {
  ["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafooo"]=>
  object(Translation_Entry)#205 (9) {
    ["is_plural"]=>
    bool(false)
    ["context"]=>
    string(255) "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    ["singular"]=>
    string(4) "fooo"
    ["plural"]=>
    NULL
    ["translations"]=>
    array(0) {
    }
    ["translator_comments"]=>
    string(0) ""
    ["extracted_comments"]=>
    string(0) ""
    ["references"]=>
    array(0) {
    }
    ["flags"]=>
    array(0) {
    }
  }
  ["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaafooo"]=>
  object(Translation_Entry)#205 (9) {
    ["is_plural"]=>
    bool(false)
    ["context"]=>
    string(255) "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
    ["singular"]=>
    string(4) "fooo"
    ["plural"]=>
    NULL
    ["translations"]=>
    array(0) {
    }
    ["translator_comments"]=>
    string(0) ""
    ["extracted_comments"]=>
    string(0) ""
    ["references"]=>
    array(0) {
    }
    ["flags"]=>
    array(0) {
    }
  }
}
```

SergeyBiryukov committed:
Add missing space after `foreach` keyword

ocean90 commented on gp-includes/things/original.php:
```diff
@@ -223,7 +235,7 @@ public function import_for_project( $project, $translations ) {
 
 		// Clear cache when the amount of strings are changed.
-		if ( $originals_added > 0 || $originals_existing > 0 || $originals_fuzzied > 0 || $originals_obsoleted > 0 ) {
+		if ( $originals_added > 0 || $originals_existing > 0 || $originals_fuzzied > 0 || $originals_obsoleted > 0 || $originals_error > 0 ) {
```
Why would we have to clear the cache if an original wasn't added?

SergeyBiryukov replied:
Good catch, fixed.

SergeyBiryukov committed:
Don't clear cache if no originals were added

ocean90 committed:
Originals: Add test for entries which exceed the maximum length for originals

ocean90 committed:
Originals: Extend test_import_for_project_with_context_which_exceeds_the_maximum_length_of_255() test

ocean90 commented on tests/tests_things/test_thing_original.php:
```diff
@@ -185,19 +185,54 @@ function test_by_project_id_and_entry_should_match_case() {
+		$this->assertEquals( 0, $originals_added );
+		$this->assertEquals( 0, $originals_existing );
+		$this->assertEquals( 0, $originals_fuzzied );
+		$this->assertEquals( 0, $originals_obsoleted );
```
This line is failing, it returns 2 for `$originals_obsoleted`.

ocean90 committed:
Originals: Fix broken `test_import_for_project_with_context_which_exceeds_the_maximum_length_of_255()` test.

> When an object is cloned, PHP 5 will perform a shallow copy of all of the object's properties. **Any properties that are references to other variables will remain references.**
— http://php.net/manual/en/language.oop5.cloning.php
